File: frontend/lib/api/accounts.ts
import { z } from "zod";
import { api } from "./client";

/**
 * Accounts API Service
 */

export const AccountSchema = z.object({
  id: z.number(),
  user_id: z.number().optional(),
  account_number: z.string().optional(),
  routing_number: z.string().nullable().optional(),
  balance: z.number(),
  reserved_balance: z.number().optional().default(0),
  currency: z.string().optional().default("USD"),
  is_main: z.boolean(),
  name: z.string(),
  masked_account_number: z.string().nullable().optional(),
});

export type Account = z.infer<typeof AccountSchema>;

export const TransactionSchema = z.object({
  id: z.string(),
  account_id: z.number(),
  merchant: z.string(),
  amount: z.number(),
  category: z.string(),
  status: z.enum(['pending', 'sent_to_kafka', 'cleared']),
  description: z.string().optional(),
  transaction_type: z.enum(['income', 'expense', 'transfer']),
  transaction_side: z.enum(['DEBIT', 'CREDIT']).optional(),
  created_at: z.string(),
});

export type Transaction = z.infer<typeof TransactionSchema>;

export const AccountSummarySchema = z.object({
  balance: z.number(),
  reserved_balance: z.number(),
  accounts: z.array(AccountSchema),
});

export type AccountSummary = z.infer<typeof AccountSummarySchema>;

export async function getAccounts(): Promise<Account[]> {
  const res = await api.get<AccountSummary>("/api/v1/accounts", {
    schema: AccountSummarySchema
  });
  return res.accounts;
}

export async function getAccountSummary(userId: number): Promise<AccountSummary> {
  return api.get<AccountSummary>(`/api/v1/accounts/${userId}`, {
    schema: AccountSummarySchema
  });
}

export async function getRecentTransactions(hours: number = 24): Promise<Transaction[]> {
  const res = await api.get<{ transactions: Transaction[] }>(`/api/v1/recent-transactions`, {
    params: { hours: hours.toString() },
    schema: z.object({ transactions: z.array(TransactionSchema) })
  });
  return res.transactions;
}

export async function getAccountTransactions(accountId: number): Promise<Transaction[]> {
  return api.get<Transaction[]>(`/api/v1/accounts/${accountId}/transactions`, {
    schema: z.array(TransactionSchema)
  });
}

File: backend/schemas/accounts.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List

class SubAccountCreate(BaseModel):
    name: str

class SubAccountRename(BaseModel):
    name: str

class InternalTransferRequest(BaseModel):
    from_account_id: int
    to_account_id: int
    amount: int
    commentary: Optional[str] = None

class AccountResponse(BaseModel):
    id: int
    name: str
    balance: int
    reserved_balance: int
    is_main: bool
    routing_number: Optional[str] = None
    masked_account_number: Optional[str] = None
    
    model_config = ConfigDict(from_attributes=True)

class BalanceResponse(BaseModel):
    balance: int
    reserved_balance: int
    user_id: int
    accounts: List[AccountResponse]

class AccountCredentialsResponse(BaseModel):
    routing_number: Optional[str]
    account_number: str
    internal_reference_id: Optional[str]
    masked_account_number: str
